src/main/kotlin/uk/gov/justice/digital/hmpps/hmppshdcapi/licences/LicenceService.kt

```diff
@@ -17,47 +15,38 @@ class LicenceService(
       throw NoDataFoundException("licence", "booking id", bookingId)
     }
 
-    val licence = licences.first().licence
+    val licenceData = licences.first().licence ?: throw NoDataFoundException("licence data", "booking id", bookingId)
 
-    if (licence.isNullOrEmpty()) {
-      throw NoDataFoundException("licence data", "booking id", bookingId)
-    }
-
-    val cas2ReferralObject = licence["bassReferral"]
-    val cas2Referral = objectMapper.convertValue(cas2ReferralObject, Cas2Referral::class.java)
-
-    val curfewObject = licence["curfew"]
-    val curfew = objectMapper.convertValue(curfewObject, Curfew::class.java)
-
-    val proposedAddressObject = licence["proposedAddress"]
-    val proposedAddress = objectMapper.convertValue(proposedAddressObject, ProposedAddress::class.java)
+    val cas2Referral = licenceData.bassReferral
+    val curfew = licenceData.curfew
+    val proposedAddress = licenceData.proposedAddress
 
     return HdcLicence(
       curfewAddress = getAddress(curfew, cas2Referral, proposedAddress),
-      firstNightCurfewHours = curfew.firstNight,
-      curfewHours = curfew.curfewHours,
+      firstNightCurfewHours = curfew?.firstNight,
+      curfewHours = curfew?.curfewHours,
     )
   }
 
-  fun getAddress(curfew: Curfew, cas2Referral: Cas2Referral, proposedAddress: ProposedAddress): CurfewAddress {
-    val isCurfewApprovedPremisesRequired = curfew.approvedPremises?.required == Decision.YES
-    val isCas2ApprovedPremisesRequired = cas2Referral.bassAreaCheck?.approvedPremisesRequiredYesNo == Decision.YES
-    val isCas2Requested = cas2Referral.bassRequest.bassRequested == Decision.YES
-    val isCas2Accepted = cas2Referral.bassOffer?.bassAccepted == OfferAccepted.YES
+  fun getAddress(curfew: Curfew?, cas2Referral: Cas2Referral?, proposedAddress: ProposedAddress?): CurfewAddress {
+    val isCurfewApprovedPremisesRequired = curfew?.approvedPremises?.required == Decision.YES
+    val isCas2ApprovedPremisesRequired = cas2Referral?.bassAreaCheck?.approvedPremisesRequiredYesNo == Decision.YES
+    val isCas2Requested = cas2Referral?.bassRequest?.bassRequested == Decision.YES
+    val isCas2Accepted = cas2Referral?.bassOffer?.bassAccepted == OfferAccepted.YES
 
     if (isCurfewApprovedPremisesRequired && !isCas2ApprovedPremisesRequired) {
-      return formatAddress(curfew.approvedPremisesAddress!!)
+      return formatAddress(curfew?.approvedPremisesAddress!!)
     }
 
     if (isCas2ApprovedPremisesRequired) {
-      return formatAddress(cas2Referral.approvedPremisesAddress!!)
+      return formatAddress(cas2Referral?.approvedPremisesAddress!!)
     }
 
     if (isCas2Requested && isCas2Accepted) {
-      val cas2Address = cas2Referral.bassOffer!!
+      val cas2Address = cas2Referral?.bassOffer!!
       return formatAddress(cas2Address)
     }
-    return formatAddress(proposedAddress.curfewAddress!!)
+    return formatAddress(proposedAddress?.curfewAddress!!)
```

Comment:
**suggestion(non-blocking):** 
This is quite a bit better than it was, maybe it'd be better to move on to the next ticket? 

you could turn this into a when: 
```kt
return when {
  isCurfewApprovedPremisesRequired && !isCas2ApprovedPremisesRequired -> 
        formatAddress(curfew?.approvedPremisesAddress!!)
  isCas2Requested && isCas2Accepted ->
        formatAddress(cas2Referral?.approvedPremisesAddress!!)
  isCas2Requested && isCas2Accepted ->
        formatAddress(cas2Referral?.bassOffer!!)
  else ->
        formatAddress(proposedAddress?.curfewAddress!!)
}
```

Still wary that although the flags are Yes | No the actual address could be null. 
That could be dealt with?: 
```kt
val address =  when {
  isCurfewApprovedPremisesRequired && !isCas2ApprovedPremisesRequired -> 
        curfew?.approvedPremisesAddress
  isCas2Requested && isCas2Accepted ->
        cas2Referral?.approvedPremisesAddress
  isCas2Requested && isCas2Accepted ->
        cas2Referral?.bassOffer
  else ->
        proposedAddress?.curfewAddress
}

return address?.let { formatAddress(it) }
```

would also be tempted to extracting out some functions to remove the flags:
```kt
val address =  when {
  useApprovedPremise(curfew, cas2Referral) -> 
        curfew?.approvedPremisesAddress
  useCas2ApprovedPremise(cas2Referral) ->
        cas2Referral?.approvedPremisesAddress
  useCas2Address(cas2Referral) ->
        cas2Referral?.bassOffer
  else ->
        proposedAddress?.curfewAddress
}

return address?.let { formatAddress(it) }
```
```